src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp

```diff
@@ -0,0 +1,731 @@
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+void SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+    SplitNode(std::vector<size_t>& points,
+              const size_t maxLeafSize,
+              const double tau,
+              const double rho)
+{
+  // We need to expand the bounds of this node properly.
+  for (size_t i = 0; i < points.size(); i++)
+    bound |= dataset->cols(points[i], points[i]);
+
+  // Calculate the furthest descendant distance.
+  furthestDescendantDistance = 0.5 * bound.Diameter();
+
+  // Now, check if we need to split at all.
+  if (points.size() <= maxLeafSize)
+  {
+    pointsIndex = new std::vector<size_t>();
+    pointsIndex->swap(points);
+    count = pointsIndex->size();
+    return; // We can't split this.
+  }
+
+  const bool split = SplitType<MetricType, MatType>::SplitSpace(bound,
+      *dataset, points, hyperplane);
+  // The node may not be always split. For instance, if all the points are the
+  // same, we can't split them.
+  if (!split)
+  {
+    pointsIndex = new std::vector<size_t>();
+    pointsIndex->swap(points);
+    count = pointsIndex->size();
+    return; // We can't split this.
+  }
+
+  std::vector<size_t> leftPoints, rightPoints;
+  // Split the node.
+  overlappingNode = SplitPoints(tau, rho, points, leftPoints, rightPoints);
+
+  // We don't need the information in points, so lets clean it.
+  std::vector<size_t>().swap(points);
+
+  // Now we will recursively split the children by calling their constructors
+  // (which perform this splitting process).
+  left = new SpillTree(this, leftPoints, tau, maxLeafSize, rho);
+  right = new SpillTree(this, rightPoints, tau, maxLeafSize, rho);
+
+  // Update count number, to represent the number of descendant points.
+  count = left->NumDescendants() + right->NumDescendants();
+
+  // Calculate parent distances for those two nodes.
+  arma::vec center, leftCenter, rightCenter;
+  Center(center);
+  left->Center(leftCenter);
+  right->Center(rightCenter);
+
+  const ElemType leftParentDistance = MetricType::Evaluate(center, leftCenter);
+  const ElemType rightParentDistance = MetricType::Evaluate(center,
+      rightCenter);
+
+  left->ParentDistance() = leftParentDistance;
+  right->ParentDistance() = rightParentDistance;
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+bool SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+    SplitPoints(const double tau,
+                const double rho,
+                const std::vector<size_t>& points,
+                std::vector<size_t>& leftPoints,
+                std::vector<size_t>& rightPoints)
```

Comment:
@rcurtin  Yes, I agree. You are right, this can make a difference for big datasets.
I have updated the code in: https://github.com/mlpack/mlpack/pull/747/commits/071588457415094e99883774af541075e3858c22
Thanks!